pkg/solana/chainwriter/chain_writer.go

```diff
@@ -0,0 +1,324 @@
+	}
+
+	// Filter derived lookup tables
+	for _, innerMap := range derivedTableMap {
+		for innerIdentifier, metas := range innerMap {
+			tableKey, err := solana.PublicKeyFromBase58(innerIdentifier)
+			if err != nil {
+				continue
+			}
+
+			// Collect public keys that are actually used
+			var usedAddresses solana.PublicKeySlice
+			for _, meta := range metas {
+				if _, exists := usedAccounts[meta.PublicKey.String()]; exists {
+					usedAddresses = append(usedAddresses, meta.PublicKey)
+				}
+			}
+
+			// Add to the filtered map if there are any used addresses
+			if len(usedAddresses) > 0 {
+				filteredLookupTables[tableKey] = usedAddresses
+			}
+		}
+	}
+
+	// Filter static lookup tables
+	for tableKey, addresses := range staticTableMap {
+		var usedAddresses solana.PublicKeySlice
+		for _, staticAddress := range addresses {
+			if _, exists := usedAccounts[staticAddress.String()]; exists {
+				usedAddresses = append(usedAddresses, staticAddress)
+			}
+		}
+
+		// Add to the filtered map if there are any used addresses
+		if len(usedAddresses) > 0 {
+			filteredLookupTables[tableKey] = usedAddresses
+		}
+	}
+
+	return filteredLookupTables
+}
+
+func (s *SolanaChainWriterService) SubmitTransaction(ctx context.Context, contractName, method string, args any, transactionID string, toAddress string, meta *types.TxMeta, value *big.Int) error {
+	programConfig, exists := s.config.Programs[contractName]
+	if !exists {
+		return fmt.Errorf("failed to find program config for contract name: %s", contractName)
+	}
+	methodConfig, exists := programConfig.Methods[method]
+	if !exists {
+		return fmt.Errorf("failed to find method config for method: %s", method)
+	}
+
+	// Configure debug ID
+	debugID := ""
+	if methodConfig.DebugIDLocation != "" {
+		var err error
+		debugID, err = GetDebugIDAtLocation(args, methodConfig.DebugIDLocation)
+		if err != nil {
+			return errorWithDebugID(fmt.Errorf("error getting debug ID from input args: %w", err), debugID)
+		}
+	}
+
+	codec := s.codecs[contractName]
+	encodedPayload, err := codec.Encode(ctx, args, method)
+	if err != nil {
+		return errorWithDebugID(fmt.Errorf("error encoding transaction payload: %w", err), debugID)
+	}
+
+	// Fetch derived and static table maps
+	derivedTableMap, staticTableMap, err := s.ResolveLookupTables(ctx, args, methodConfig.LookupTables)
+	if err != nil {
+		return errorWithDebugID(fmt.Errorf("error getting lookup tables: %w", err), debugID)
+	}
+
+	// Resolve account metas
+	accounts, err := GetAddresses(ctx, args, methodConfig.Accounts, derivedTableMap, s.reader)
+	if err != nil {
+		return errorWithDebugID(fmt.Errorf("error resolving account addresses: %w", err), debugID)
+	}
+
+	// Filter the lookup table addresses based on which accounts are actually used
+	filteredLookupTableMap := s.FilterLookupTableAddresses(accounts, derivedTableMap, staticTableMap)
+
+	// Fetch latest blockhash
+	blockhash, err := s.reader.LatestBlockhash(ctx)
```

Comment:
Completely agree

Reply:
> In my view, fetching and setting the latest blockHash should be entirely encapsulated inside the TXM.
> No point in letting a caller of TXM setting it.

We discussed this with Amit but chose not to move forward with that approach yet because there are a few places in the codebase where the caller needs the blockhash to do some additional operations before passing the transaction on to the TXM. 

An example is [this code](https://github.com/smartcontractkit/chainlink-solana/blob/e2ff830d9bc57ada79ba9b2a9e8304c4f6f6b816/pkg/solana/chain.go#L552-L588)

> For one, the caller has no idea when will the TXM actually process this Tx. It will go into a queue, and get picked up when TXM is ready, which may be pretty late, causing this blockHash to get stale.

Agreed. Having the TXM itself fetch and assign the latest blockhash would reduce the risk of staleness and hide complexity from the caller. I'm okay refactoring these other call sites so that the TXM becomes the single source of truth for fetching the current blockhash. That refactor isn’t difficult, but it does require some care to ensure no existing logic depends on a blockhash before calling the TXM. 

@amit-momin if you are okay with this, I can create a ticket inside the [optimization track](https://smartcontract-it.atlassian.net/browse/NONEVM-1023) to address this.

> Secondly, I think the tXM anyways sets recent blockhashes on all txs that it processes. So this is duplicate.

Currently, the TXM only fetches and updates the blockhash when it needs to rebroadcast expired transactions ([see code](https://github.com/smartcontractkit/chainlink-solana/pull/github.com/smartcontractkit/chainlink-solana/blob/e2ff830d9bc57ada79ba9b2a9e8304c4f6f6b816/pkg/solana/txm/txm.go#L569-L577))

Reply:
An in-between solution could be making the blockhash an optional parameter?

Reply:
I don’t think this issue blocks the current Chain Writer PR and it should be merged as-is with the existing blockhash call. 

If we decide to optimize later, I’d suggest going straight to a long term solution where the TXM manages blockhash fetching and assignment entirely, rather than an in-between approach if possible. 

In doing so, we’d refactor the callers so they no longer fetch the blockhash themselves (potentially through the Chain Writer rather than a direct Enqueue call 💪 ). In the case a caller really needs the blockhash, the TXM could simply overwrite it later. This would still ensure the correctness of the flow, avoiding the special-case complexity. The tradeoff is that we would be making a duplicate blockhash call for that particular case if needed.